### ISSUE TITLE
Support for Python Wheels?

### ISSUE THREAD

**Issue by data-henrik** [ibmdb/python-ibmdb#601]
It would be great to have some form of binary distribution of this Db2 driver for Python. It would simplify building Docker containers and help reducing the container size and setup time.

Any plans? Should this be an Aha feedback?

**Comment by amukherjee28** [comment]
Hi Henrik,

Can you please elaborate a bit more on what exactly are you suggesting in here, when you say binary distribution of python driver.

If I guess correctly are talking about having a docker image of ibm_db driver into it so that when we spin up the docker image it will have the setup ready for python to work on db2 servers.

The docker image should have the complete setup in place with the latest available package.

Is that what you meant?

Feel free to correct me if my understanding is wrong.

Thanks.

**Comment by data-henrik** [comment]
I was referring to https://packaging.python.org/discussions/wheel-vs-egg/

ibm_db requires to download the ODBC / CLI driver and some compilation. Ideally, everything is already ready and could directly used. At least on the popular platforms.

The request is not to provide a Docker container, but to simplify the use of the Python driver. One such use case would be a dockerized Python app.

**Comment by amukherjee28** [comment]
Hi Henrik,

Thank you for the explanation here.

I think now I have a good understanding of what you mean here.

Let me go through this request once and get back to you at the earliest.

Thanks.

**Comment by bimalkjha** [comment]
@data-henrik Good to have an Aha entry for this enhancement to prioritize it. Please post the Aha number here. Thanks.

**Comment by data-henrik** [comment]
See DB2CON-I-66

**Comment by bimalkjha** [comment]
Opened Jira Epic: https://jira.rocketsoftware.com/browse/DBC-11180

**Comment by zinal** [comment]
The Aha Idea page is here: [https://ibm-data-and-ai.ideas.ibm.com/ideas/DB2CON-I-66](https://ibm-data-and-ai.ideas.ibm.com/ideas/DB2CON-I-66)

**Comment by zinal** [comment]
Support for Python Wheels generally means having a hosted set of binary packages, which should probably include ODBC/CLI package (maybe in a partial/stripped form) and a pre-compiled Python driver.
Having the pre-built packages for popular platforms (like Ubuntu, RHEL, CoreOS) will enable faster and more stable installation, and smaller Docker images (not bloated with source code, compilation data and unneeded Db2 files).

**Comment by c-muncan** [comment]
this would be extremely beneficial for portability on native z/OS where the XL C compiler is optional but for ibm_db, it is a requirement

**Comment by Earammak** [comment]
@c-muncan As part of this EPIC, we have a plan of supporting wheels for popular platforms only [ linux , macos and windows].
Please refer to (https://pythonwheels.com/).

**Comment by powellquiring** [comment]
I think I just ran into this issue when trying to create a code engine python application on ibm cloud.  I think I captured all the relevant logs:
```
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 11.2 MB/s eta 0:00:00
Collecting python-dotenv (from -r requirements.txt (line 5))
Downloading python_dotenv-1.0.0-py3-none-any.whl (19 kB)
Collecting Flask_SQLAlchemy==2.5.1 (from -r requirements.txt (line 6))
Downloading Flask_SQLAlchemy-2.5.1-py2.py3-none-any.whl (17 kB)
Collecting SQLAlchemy==1.3.22 (from -r requirements.txt (line 7))
Downloading SQLAlchemy-1.3.22.tar.gz (6.3 MB)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 64.4 MB/s eta 0:00:00
Installing build dependencies: started
Installing build dependencies: finished with status 'done'
Getting requirements to build wheel: started
Getting requirements to build wheel: finished with status 'done'
Preparing metadata (pyproject.toml): started
Preparing metadata (pyproject.toml): finished with status 'done'
Collecting ibm-db-sa (from -r requirements.txt (line 8))
Downloading ibm_db_sa-0.4.0-py3-none-any.whl (31 kB)
Collecting ibm-db (from -r requirements.txt (line 9))
Downloading ibm_db-3.1.4.tar.gz (1.4 MB)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 62.3 MB/s eta 0:00:00
Installing build dependencies: started
Installing build dependencies: finished with status 'done'
Getting requirements to build wheel: started
Getting requirements to build wheel: finished with status 'error'
error: subprocess-exited-with-error
slackbot-backend-run-230626-10465945-47q69-pod/step-build-and-push:
× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> [8 lines of output]
Detected 64-bit Python
Detected platform = linux, uname = x86_64
Downloading https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64_odbc_cli.tar.gz
Downloading DSDriver from url =  https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/linuxx64_odbc_cli.tar.gz
Pre-requisite check [gcc] : Passed
slackbot-backend-run-230626-10465945-47q69-pod/step-build-and-push:
No Python.h header file detected.
Please install python-devel or python3-devel as instructed in "https://github.com/ibmdb/python-ibmdb/blob/master/README.md#KnownIssues" and continue with the installation
[end of output]
slackbot-backend-run-230626-10465945-47q69-pod/step-build-and-push:
```

I'm kind of stuck.  The complaint is a Python.h file is needed, https://github.com/ibmdb/python-ibmdb/blob/master/README.md#KnownIssues.  Is this to build the ibm db2 driver?  Could this error be eliminated if the ibm_db package was a wheel?

Note I'm a step removed from the docker image build process that is taking my python source code and turning it into a docker container image.  The container is not being built on my desktop.  The docker image is being built by a third party (IBM Code Engine) and I think it is purposefully leaving out the details - so I'm finding it difficult to troubleshoot.

**Comment by bimalkjha** [comment]
@powellquiring The python.h file comes with python3-dev package and adding steps in docker file to install this OS package before installing ibm_db should fix the issue. We need python.h to compile native code of ibm_db during installation and yes, having wheel image will solve this issue. You got the correct doc link for this issue.
We are working on providing wheel support for wheel supported platforms and it should be available with next release of ibm_db. Thanks.

**Comment by data-henrik** [comment]
what's the ETA for wheel support?

**Comment by Earammak** [comment]
@data-henrik, We are working on it. We will soon release a new version with wheels.

Thanks

**Comment by powellquiring** [comment]
What's the ETA for wheel support?

**Comment by bimalkjha** [comment]
Wheel Support code changes got merged to master branch using PR: https://github.com/ibmdb/python-ibmdb/pull/862

**Comment by data-henrik** [comment]
What are the next steps and the timeline to really have wheel support? Or was I too early to test it?

**Comment by bimalkjha** [comment]
@data-henrik We have just released a new version of ibm_db v3.2.0 with Wheel support. Please install the latest version of ibm_db and test it.
Hence, now closing this issue. Thanks.

**Comment by data-henrik** [comment]
> @data-henrik @powellquiring We have just released a new version of ibm_db v3.2.0 with Wheel support. Please install the latest version of ibm_db and test it. Hence, now closing this issue. Thanks.

installation works, I need to run tests - thank you
